byoqm/models/code_climate.py

```diff
@@ -22,241 +17,76 @@ def __init__(self):
-            "duplication": self.duplication,
-            "lines of code": self.file_length,
-            "method length": self.method_length,
-            "return statements": self.return_statements,
-            "identical blocks of code": self.identical_blocks_of_code,
-            "argument count": self.argument_count,
-            "method count": self.method_count,
-            "complex logic": self.complex_logic,
+            "metrics": {
+                "method_length": "./metrics/method_length.py",
+                "file_length": "./metrics/file_length.py",
+                "argument_count": "./metrics/argument_count.py",
+                "complex_logic": "./metrics/complex_logic.py",
+                "method_count": "./metrics/method_count.py",
+                "return_statements": "./metrics/return_statements.py",
+                "identical_blocks_of_code": "./metrics/identical_codeblocks.py",
+                "similar_blocks_of_code": "./metrics/similar_codeblocks.py",
+                "nested_controlflows": "./metrics/nested_controlflows.py",
+            },
+            "aggregations": {
+                "quality": self.maintainability,
+                "maintainability": self.maintainability,
+                "duplication": self.duplication,
+                "cognitive complexity": self.cognitive_complexity,
+                "structural issues": self.structural_issues,
+                "cyclomatic complexity": self.cyclomatic_complexity,
+            },
         }
         return model
 
-    def save_to_csv(self, path="./output"):
-        file_location = path + "/" + str(date.today()) + ".csv"
-        if not os.path.exists(path):
-            os.mkdir(path)
-        with open(file_location, "w") as file:
-            writer = csv.writer(file)
-            writer.writerow(["Metric", "Value"])
-            for metric, func in self.getDesc().items():
-                if metric == "identical blocks of code":
-                    writer.writerow([metric, func(35)])
-                else:
-                    writer.writerow([metric, func()])
+    def get_aggregated_results(self):
+        return self.maintainability()
 
     def maintainability(self):
-        return 7 + self.duplication()
+        return (
+            self.duplication()
+            + self.cognitive_complexity()
+            + self.structural_issues()
+            + self.cyclomatic_complexity()
+        )
 
     def duplication(self) -> int | float:
-        return self.identical_blocks_of_code(35) + self.similar_blocks_of_code()
-
-    def cognitive_complexity(self):
-        pass
+        with self.results.open() as f:
+            reader = csv.DictReader(f)
+            sum = 0
+            for row in reader:
+                if row["Metric"] == "identical_blocks_of_code":
+                    sum += int(row["Value"])
+                if row["Metric"] == "similar_blocks_of_code":
+                    sum += int(row["Value"])
+            return sum
```

Comment:
Should the aggregation of underlying metrics really be combined inside of the quality model or is it better to move it outside? I cannot recall what conclusion we came to when discussing the scope of quality model definition files.

Reply:
We have talked about a quality model being a "set of aggregation methods", so that is what I am trying to reflect. I do see your concerns though, as now we have some kind of "double coupling" where the full results are only available after having run metrics through `main.py` and the aggregation methods in the model.